crates/project_panel/src/project_panel.rs

```diff
@@ -1304,46 +1293,101 @@ impl ProjectPanel {
-                if clipboard_entries.is_cut() {
-                    self.project
-                        .update(cx, |project, cx| {
-                            project.rename_entry(clipboard_entry.entry_id, new_path, cx)
-                        })
-                        .detach_and_log_err(cx);
+                let clip_entry_id = clipboard_entry.entry_id;
+                let is_same_worktree = clipboard_entry.worktree_id == worktree_id;
+                let relative_worktree_source_path = if !is_same_worktree {
+                    let target_base_path = worktree.read(cx).abs_path();
+                    let clipboard_project_path =
+                        self.project.read(cx).path_for_entry(clip_entry_id, cx)?;
+                    let clipboard_abs_path = self
+                        .project
+                        .read(cx)
+                        .absolute_path(&clipboard_project_path, cx)?;
+                    Some(relativize_path(
+                        &target_base_path,
+                        clipboard_abs_path.as_path(),
+                    ))
+                } else {
+                    None
+                };
+                let task = if clip_is_cut && is_same_worktree {
+                    let task = self.project.update(cx, |project, cx| {
+                        project.rename_entry(clip_entry_id, new_path, cx)
+                    });
+                    PasteTask::Rename(task)
                 } else {
+                    let entry_id = if is_same_worktree {
+                        clip_entry_id
+                    } else {
+                        entry.id
+                    };
                     let task = self.project.update(cx, |project, cx| {
-                        project.copy_entry(clipboard_entry.entry_id, new_path, cx)
+                        project.copy_entry(entry_id, relative_worktree_source_path, new_path, cx)
                     });
-                    tasks.push(task);
-                }
+                    PasteTask::Copy(task)
+                };
+                let need_delete = !is_same_worktree && clip_is_cut;
+                paste_entry_tasks.insert((clip_entry_id, need_delete), task);
             }
 
             cx.spawn(|project_panel, mut cx| async move {
-                let entry_ids = futures::future::join_all(tasks).await;
-                if let Some(Some(entry)) = entry_ids
-                    .into_iter()
-                    .rev()
-                    .find_map(|entry_id| entry_id.ok())
-                {
+                let mut last_succeed = None;
+                let mut need_delete_ids = Vec::new();
+                for ((entry_id, need_delete), task) in paste_entry_tasks.into_iter() {
+                    match task {
+                        PasteTask::Rename(task) => {
+                            let result = task.await;
+                            if let Some(CreatedEntry::Included(entry)) = result.ok() {
```

Comment:
```suggestion
                            if let Some(CreatedEntry::Included(entry)) = task.await.log_err() {
```

Same below — let's log any task error as otherwise we might miss that.

Also we can rename `need_delete` into more grammatically correct `needs_delete`.


----------------

To double check: I see that rename tasks do not check for `need_delete`. 
Is that true that even in the new, special case you work on that is not needed?

Reply:
> Is that true that even in the new, special case you work on that is not needed?

yes, this don't need delete.